### ISSUE TITLE
quinn-proto API clarifications

### ISSUE THREAD

**Issue by infinity0** [quinn-rs/quinn#924]
Hi, I am looking at the quinn-proto API docs. [Endpoint](https://docs.rs/quinn-proto/0.6.1/quinn_proto/generic/struct.Endpoint.html) is quite simple but [Connection](https://docs.rs/quinn-proto/0.6.1/quinn_proto/generic/struct.Connection.html) is more complex, and so I have the following questions for clarification: 

- Can you clarify when `poll_endpoint_events` is supposed to be called? All the other `poll_*` methods have docs describing this.
- Where it says "the application performed some I/O on the connection" can you:
  - Please explicitly list out which of the other functions count as "I/O"? I am guessing anything that might change the state of the state machine, however it would be best to have this mentioned explicitly. For example this might be unclear for `reset`. I am guessing this may change some timeouts, but someone less familiar with networking might not realise this counts as "I/O" especially since the docs says "abandon sending" which sounds like the lack of I/O.
  - Having to call `poll_*` afterwards, this makes sense for commands that might want to write onto the network, since you are asking the state machine to do more things. However for read commands, this is less obvious. I'm consuming things that have already happened, why should anything timeout or transmit in response to this? At best, I am emptying a buffer inside the state, then this might mean that future incoming network packets can now trigger additional things to transmit (i.e. I can now ACK it because there's more space in the buffer), but as a direct response to the read command I'd intuitively expect this not to be the case. Can you clarify if the documentation should be changed to say "performed some write on the connection" instead of "I/O"?
- If `poll_timeout` returns 1000 then later 500, does that cancel the original timer at 1000? Or are both 500 and 1000 set?

**Comment by infinity0** [comment]
Perhaps "performed some write" is not the best wording either since it sounds like it excludes "reset". AIUI some functions (including the send functions) are commands produced by another component **into** the state machine, some functions (including the recv functions) are taking results **out of** the state machine and consuming it. Conceptually you could write a single function

```rust
handle_every_input(
    &mut self,
    { ConnectionEvent | Instant | Send | Ping | Reset | ... } input
) -> Vec<{ Instant | Event | EndpointEvent | Recv | ... }>
```

where by `A | B` I mean some enum type consisting of A and B, and this function would encapsulate the whole protocol logic of `Connection`.

**Comment by Ralith** [comment]
Thanks for the input! These seem like good suggestions overall.

> Can you clarify if the documentation should be changed to say "performed some write on the connection" instead of "I/O"?

The documentation is correct as written, because reading can cause the receiver to issue flow control updates to the sender to notify it of the newly available buffer space.

**Comment by djc** [comment]
> If `poll_timeout` returns 1000 then later 500, does that cancel the original timer at 1000? Or are both 500 and 1000 set?

There is no concept of cancellation here: `poll_timeout()` simply returns the earliest time at which the `Connection` has something to do (unless other external events come up sooner, of course). So if it returns 1000 and then 500, you should just make sure to poll the connection for events after 500 (and make sure to call `poll_timeout()` again after that, of course -- in which case you might get another 500 originating from what was previously 1000).

Here's roughly what the [`quinn::ConnectionDriver` loop](https://github.com/quinn-rs/quinn/blob/main/quinn/src/connection.rs#L253) does:

* Call `handle_event()` for each event from the `Endpoint`
* Pull transmissions out of the `Connection` through `poll_transmit()`
* Set a timer based on the result of `poll_timeout()`
* Pull endpoint events out of the `Connection` through `poll_endpoint_events()`
* Pull application events out of the `Connection` through `poll()`

I think we have roughly 4 types of methods on `Connection`:

* Simple getters (taking `&self`)
* Incoming event handlers (`handle_event()` and `handle_timeout()`)
* Polling functions (`poll()`, `poll_timeout()`, `poll_endpoint_events()`)
* State machine mutators (other methods taking `&mut self`) -- I think these are what we mean by "performing I/O"

Basically, any time after you call an event handler or a state machine mutator, you should poll all three of the polling functions to make sure the application/networking state is up-to-date according to the state machine.

Let me know if that helps -- otherwise we'd be happy to field further questions. It would be great if you can turn your newly acquired understanding in a docs PR, since you have the best idea of what docs are lacking and what kinds of descriptions would help.

As for `someone less familiar with networking`, I think we would not expect these types of users to use the quinn-proto API, but use the quinn API instead. It would be interesting for us to understand why you decided to go with the lower-level API rather than the higher-level one!

**Comment by infinity0** [comment]
> As for `someone less familiar with networking`, I think we would not expect these types of users to use the quinn-proto API, but use the quinn API instead. It would be interesting for us to understand why you decided to go with the lower-level API rather than the higher-level one!

I was just giving an example on someone who might be confused with the docs. I myself can guess with some confidence which ones count as "I/O" but it still would be nice to have them explicitly stated in the docs, it helps to decrease the mental burden required in using the API, and the potential for nasty surprises later on.

As for why - generally I agree with the principle of implementing core protocol logic as a deterministic state machine. To fully realise this principle, the API should also be easy to use - otherwise every consumer of the API has to implement a lot of similar logic, which partially defeats the whole point of implementing the core logic purely - it does not capture the whole logic of the protocol, there is some common external part that must always be reimplemented. Sometimes depending on the language it's impossible to factor away this common logic into a reusable part, but then the next best thing is to describe it in the docs as fully as possible.

More concretely, I am writing an experimental Haskell framework for deterministic protocol implementations, and the single-handler API is what I settled on (analogous to `handle_every_input` I wrote above). Once you have this simple API, in Haskell you even have the power to compose different protocols together (e.g. stacking multiple layers) in a general way, using only this API and without any knowledge of the protocol specifics. To add QUIC support to that, I was looking into how feasible it would be to create Haskell-Rust bindings to this library, then implementing the simple API wrapper around that. (There is a Haskell QUIC implementation WIP but it is strongly coupled to network IO - https://github.com/kazu-yamamoto/quic/issues/11.)

**Comment by infinity0** [comment]
> There is no concept of cancellation here: poll_timeout() simply returns the earliest time at which the Connection has something to do (unless other external events come up sooner, of course).

Great thanks for that. To be clear, `handle_timeout()` would NOP if it got a spurious `Instant` signal at which there was no timeout, or does it depend on it being given correctly?

> Basically, any time after you call an event handler or a state machine mutator, you should poll all three of the polling functions to make sure the application/networking state is up-to-date according to the state machine.

There are actually 4 :) `poll` and `poll_{timeout,endpoint_events,transmit}`. It is indeed the "safest" option for consumers of this API to poll all of them, but I thought perhaps you intended to imply it's possible to make some optimisations by omitting polling some of them. If it's possible this might change later on though, then it would be best to advise the safest option for consumers of the API. I assume it shouldn't matter the order in which they are polled, that is entirely up to the consumer?

Another question - suppose I have a `ConnectEvent` with `now: 500`, would it cause problems if I accidentally give this to the state machine *after* I give a timeout event of `501`? Or even `500`?

> It would be great if you can turn your newly acquired understanding in a docs PR, since you have the best idea of what docs are lacking and what kinds of descriptions would help.

Sure I'll have a go.

**Comment by Ralith** [comment]
> To be clear, handle_timeout() would NOP if it got a spurious Instant signal at which there was no timeout, or does it depend on it being given correctly?

`handle_timeout` just processes all timeouts which are expired wrt. the passed-in `now`, so it is robust to spurious polling.

**Comment by djc** [comment]
> Another question - suppose I have a `ConnectEvent` with `now: 500`, would it cause problems if I accidentally give this to the state machine after I give a timeout event of `501`? Or even `500`?

If that caused severe issues we would probably treat it as a bug, but of course we do generally expect monotonically increasing time.

>  I assume it shouldn't matter the order in which they are polled, that is entirely up to the consumer?

It doesn't matter much, no. Of course some things depend on the application, for example, for most applications it probably makes sense to prioritize transmissions over handling endpoint events.
